File: slash/src/commands/movie.ts
import fetch from "node-fetch";
import {
  CommandContext,
  CommandOptionType,
  InteractionResponseFlags,
  SlashCommand,
  SlashCreator,
} from "slash-create";

interface TMDBMovie {
  adult: boolean;
  backdrop_path: string | null;
  belongs_to_collection: object | null;
  budget: number;
  genres: { id: number; name: string }[];
  homepage: string | null;
  id: number;
  imdb_id: string | null;
  original_language: string;
  original_title: string;
  overview: string | null;
  popularity: number;
  poster_path: string | null;
  production_companies: { name: string }[];
  release_date: string;
  revenue: number;
  runtime: number;
  tagline: string | null;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}

export default class extends SlashCommand {
  constructor(creator: SlashCreator) {
    super(creator, {
      name: "movie",
      description: "Search for your favorite movie online.",
      options: [
        {
          name: "title",
          description: "The title of the movie you're looking for.",
          type: CommandOptionType.STRING,
          required: true,
        },
        {
          name: "year",
          description: "The year the movie was released.",
          type: CommandOptionType.INTEGER,
        },
      ],
    });
  }

  async run(ctx: CommandContext) {
    // Create querystring params for request
    const params = new URLSearchParams({
      api_key: process.env.TMDB_KEY ?? "",
      query: ctx.options.title,
    });
    if (ctx.options.year) params.set("year", String(ctx.options.year));

    // Find movie
    const res = await fetch(
      `https://api.themoviedb.org/3/search/movie?${params.toString()}`
    );
    const body: { results: { id: number }[] } = await res.json();

    // Send message if no movie was found.
    if (!body.results.length) {
      return ctx.send(
        "There's no movie by that name (in that year, if specified).",
        {
          flags: InteractionResponseFlags.EPHEMERAL,
        }
      );
    }

    await ctx.defer();

    const movieRes = await fetch(
      `https://api.themoviedb.org/3/movie/${body.results[0].id}?api_key=${process.env.TMDB_KEY}`
    );
    const movie: TMDBMovie = await movieRes.json();

    return ctx.editOriginal({
      embeds: [
        {
          title: `${movie.title} (${movie.release_date.slice(0, 4)})`,
          url: `https://www.imdb.com/title/${movie.imdb_id}`,
          description: movie.overview ?? "",
          fields: [
            {
              name: "Genres",
              value: movie.genres.map((g) => g.name).join(", "),
              inline: true,
            },
            {
              name: "Length",
              value: `${movie.runtime} min`,
              inline: true,
            },
            {
              name: "Box Office",
              value: movie.revenue.toLocaleString("en-US", {
                style: "currency",
                currency: "USD",
              }),
              inline: true,
            },
          ],
          image: {
            url: `https://image.tmdb.org/t/p/w500${movie.backdrop_path}`,
          },
          thumbnail: {
            url: `https://image.tmdb.org/t/p/w500${movie.poster_path}`,
          },
        },
      ],
    });
  }
}

File: backend/commands/movie.go
package commands

import (
	"backend/types"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"strings"
)

var tmdb_key = os.Getenv("TMDB_KEY")

type SearchResult struct {
	Page    int `json:"page"`
	Results []struct {
		ID int `json:"id"`
	} `json:"results"`
	TotalResults int `json:"total_results"`
	TotalPages   int `json:"total_pages"`
}

type Movie struct {
	Adult        bool   `json:"adult"`
	BackdropPath string `json:"backdrop_path"`
	Genres       []struct {
		ID   int    `json:"id"`
		Name string `json:"name"`
	} `json:"genres"`
	ID               int     `json:"id"`
	ImdbID           string  `json:"imdb_id"`
	OriginalLanguage string  `json:"original_language"`
	OriginalTitle    string  `json:"original_title"`
	Overview         string  `json:"overview"`
	Popularity       float64 `json:"popularity"`
	PosterPath       string  `json:"poster_path,omitempty"`
	ReleaseDate      string  `json:"release_date"`
	Revenue          int     `json:"revenue"`
	Runtime          int     `json:"runtime"`
	Tagline          string  `json:"tagline"`
	Title            string  `json:"title"`
}

func movie(i *types.Interaction, data *types.ApplicationCommandInteractionData) {
	title, _ := data.GetOption("title")
	year, noYear := data.GetOption("year")

	// Format the URL
	query := url.QueryEscape(title.StringValue())
	var searchURL string
	if noYear != nil {
		searchURL = fmt.Sprintf("https://api.themoviedb.org/3/search/movie?api_key=%s&query=%s", tmdb_key, query)
	} else {
		searchURL = fmt.Sprintf("https://api.themoviedb.org/3/search/movie?api_key=%s&query=%s&year=%f", tmdb_key, query, year.IntValue())
	}

	// Search for movie
	searchResp, err := http.Get(searchURL)
	if err != nil {
		fmt.Println(err)
		i.Error("No movie found by that name")
		return
	}

	// Get search data
	defer searchResp.Body.Close()
	var searchData SearchResult
	if err = json.NewDecoder(searchResp.Body).Decode(&searchData); err != nil {
		fmt.Println(err)
		i.Error("Error searching for that movie.")
		return
	}

	// Show error if no movie is found
	if searchData.TotalResults == 0 {
		i.Error("I didn't find any movie by that name.")
		return
	}

	// Retrieve movie
	movieResp, err := http.Get(fmt.Sprintf("https://api.themoviedb.org/3/movie/%d?api_key=%s&language=%s", searchData.Results[0].ID, tmdb_key, i.GuildLocale))
	if err != nil {
		fmt.Println(err)
		i.Error("Error finding details about that movie.")
		return
	}

	// Get movie data
	defer movieResp.Body.Close()
	var movie Movie
	if err = json.NewDecoder(movieResp.Body).Decode(&movie); err != nil {
		fmt.Println(err)
		i.Error("Error finding details about that movie.")
		return
	}

	// Respond with movie data
	i.Respond(types.InteractionResponse{
		Type: types.ResponseChannelMessageWithSource,
		Data: types.ResponseData{
			Embeds: []*types.Embed{{
				Title:       fmt.Sprintf("%s (%s)", movie.Title, movie.ReleaseDate[:4]),
				URL:         "https://letterboxd.com/film/" + KebabCaseTitle(movie.OriginalTitle),
				Description: movie.Overview,
				Fields: []types.EmbedField{
					{
						Name: "Genres",
						Value: func() string {
							res := ""
							for _, genre := range movie.Genres {
								res += genre.Name + ", "
							}
							return res[:len(res)-2]
						}(),
						Inline: true,
					},
					{
						Name:   "Length",
						Value:  fmt.Sprintf("%d min", movie.Runtime),
						Inline: true,
					},
				},
				Image: types.EmbedImage{
					URL: "https://image.tmdb.org/t/p/w500" + movie.BackdropPath,
				},
				Thumbnail: types.EmbedImage{
					URL: "https://image.tmdb.org/t/p/w500" + movie.PosterPath,
				},
			}},
			Components: []interface{}{
				types.ActionRow{
					Type: types.ComponentActionRow,
					Components: []interface{}{
						types.Button{
							Type:  types.ComponentButton,
							Style: types.ButtonLink,
							URL:   fmt.Sprintf("https://www.themoviedb.org/movie/%d-%s/watch", movie.ID, KebabCaseTitle(movie.Title)),
							Label: "Watch Now",
						},
					},
				},
			},
		},
	})
}

func KebabCaseTitle(title string) string {
	return strings.ReplaceAll(strings.ToLower(title), " ", "-")
}
